af_ktls.c

```diff
@@ -358,30 +379,569 @@ static void tls_update_senpage_ctx(struct tls_sock *tsk, size_t size)
+	/*
+	 * Possible security vulnerability since skb can require
+	 * more than ALG_MAX_PAGES sg's.
+	 * TODO: So what exactly happens if skb_to_sgvec causes more
+	 * than ALG_MAX_PAGES fragments? Consider allocating kernel
+	 * pages
+	 * tls_read_size already copied headers and aad. Therefore
+	 * this simply needs to pass the encrypted data + message
+	 */
+	nsg = skb_to_sgvec(skb, &tsk->sgin[1], rxm->offset +
+			prepend,
+			rxm->full_len - prepend);
+	tls_make_aad(tsk, 1, tsk->aad_recv,
+			rxm->full_len - overhead,
+			tsk->iv_recv);
+
+	/*
+	 * Decrypt in place.
+	 * After this function call, the decrypted data will be in
+	 * rxm->offset. We must therefore account for the fact that
+	 * the lengths of skbuff_in and skbuff_out are different
+	 */
+
+	ret = tls_do_decryption(tsk,
+			tsk->sgin,
+			tsk->sgin,
+			rxm->full_len - overhead);
+
+	if (ret < 0)
+		goto decryption_fail;
+
+
+	ret = tls_post_process(tsk, skb);
+
+	if (ret < 0)
+		goto decryption_fail;
+
+	return 0;
+decryption_fail:
+	return ret;
+}
+/*
+ * Returns the length of the unencrypted message, plus overhead
+ * Note that this function also populates tsk->header which is later
+ * used for decryption
+ * TODO: Revisit
+ * In TLS we automatically bail if we see a non-TLS message. In DTLS
+ * we should determine if we got a corrupted message vs a control msg
+ * Right now if the TLS magic bit got corrupted it would incorrectly
+ * misinterpret it as a non-TLS message
+ * Returns 0 if more data is necessary to determine length
+ * Returns <0 if error occured
+ */
+static inline ssize_t tls_read_size(struct tls_sock *tsk, struct sk_buff *skb)
+{
+	int ret;
+	size_t data_len = 0;
+	size_t datagram_len;
+	size_t prepend;
+	char first_byte;
+	char *header;
+	struct tls_rx_msg *rxm;
+
+	prepend = IS_TLS(tsk) ? KTLS_TLS_PREPEND_SIZE : KTLS_DTLS_PREPEND_SIZE;
+	header = tsk->header_recv;
+	xprintk("--> %s", __func__);
+
+	rxm = tls_rx_msg(skb);
+
+	ret = skb_copy_bits(skb, rxm->offset, &first_byte, 1);
+	if (ret < 0)
+		goto read_failure;
+
+	/*Check the first byte to see if its a TLS record */
+	if (first_byte != KTLS_RECORD_DATA) {
+		ret = -EBADMSG;
+		goto read_failure;
+	}
+
+	/*
+	 *We have a TLS record. Check that msglen is long enough to read
+	 *the length of record.
+	 *We must not check this before checking the first byte, since
+	 *that will cause unencrypted
+	 *messages shorter than KTLS_TLS_PREPEND_SIZE to not be read
+	 */
+	if (rxm->offset + prepend > skb->len) {
+		ret = 0;
+		goto read_failure;
+	}
+
+	/*
+	 *Copy header to use later in decryption.
+	 *An optimization could be to zero-copy, but you'd
+	 *have to be able to deal with frag_list bs. This function call
+	 *takes care of that
+	 *Overhead is relatively small (13 bytes for TLS, 21 for DTLS)
+	 */
+	ret = skb_copy_bits(skb, rxm->offset, tsk->header_recv,
+			prepend);
+
+	if (ret < 0)
+		goto read_failure;
+
+	if (IS_TLS(tsk)) {
+		data_len = ((header[4] & 0xFF) | (header[3] << 8));
```

Comment:
This was taken from original sources. We should encapsulate in a macro probably. The same for else branch.